# Assignment A1 - Data import

Course "Data processing and Visualization", IE500417, NTNU, Autumn, 2025

https://www.ntnu.edu/studies/courses/IE500417

## Task description

This assignment consists of several parts:
1. Set up development environment on your computer. It includes installation and setup of tools which will be necessary in other course assignments. 
2. Load data sets of several popular data formats: CSV, JSON, XML
3. Do simple checks and analysis of loaded data.

## Submission details

The assignment must be handed in on Blackboard. The following must be handed in:
1. Report in HTML or PDF format describing the results of this assignment. Preferably, it is generated from the Jupyter notebook you used (Hint: In Jupyter: File > Download as > HTML). Alternatively (if you use plain Python or other tools), prepare a readable report that contains figures and source code snippets necessary to understand your work.
2. Source code that you used to generate the results. This could be the the Jupyter notebook file, python source files, Matlab files, etc.

Deadlines and grading information on Blackboard.

---
## Detailed steps: Part 2
---


### Part 2.1: Import data from CSV file

First you must import the Pandas library. It tells the Python interpreter, what functions are available in the Pandas. to make the writing shorter, one typically loads the pandas library with an alias pd. (The following block is python source code.)

In [1]:
# import all functions from Pandas library, they will be accessible through alias pd
import pandas as pd; 

# for example the read_csv function will be accessible as pd.read_csv(...)

Now when the Pandas library is imported, load data from the CSV file into a Python DataFrame object called `houses`. Use the .CSV file attached to this assignment: real_estate.csv .
(Data source: https://support.spatialkey.com/spatialkey-sample-csv-data/ )
It contains data about sold real estate (apartment, houses, etc).

**Pay attention to the texts below. Check out all the blocks where you have to insert code!**

In [2]:
# Write code to load the CSV data into a Pandas Dataframe
houses = pd.read_csv("Data/real_estate.csv") 

Print the first 5 rows of the dataframe.

In [3]:
# Your code here

houses.head()

street        city    zip state  beds  baths  sq__ft  \
0      3526 HIGH ST  SACRAMENTO  95838    CA     2      1     836   
1       51 OMAHA CT  SACRAMENTO  95823    CA     3      1    1167   
2    2796 BRANCH ST  SACRAMENTO  95815    CA     2      1     796   
3  2805 JANETTE WAY  SACRAMENTO  95815    CA     2      1     852   
4   6001 MCMAHON DR  SACRAMENTO  95824    CA     2      1     797   

          type                     sale_date  price   latitude   longitude  
0  Residential  Wed May 21 00:00:00 EDT 2008  59222  38.631913 -121.434879  
1  Residential  Wed May 21 00:00:00 EDT 2008  68212  38.478902 -121.431028  
2  Residential  Wed May 21 00:00:00 EDT 2008  68880  38.618305 -121.443839  
3  Residential  Wed May 21 00:00:00 EDT 2008  69307  38.616835 -121.439146  
4  Residential  Wed May 21 00:00:00 EDT 2008  81900  38.519470 -121.435768

Print the last 5 rows of the dataframe.

In [4]:
# Your code here

houses.tail()

street             city    zip state  beds  baths  sq__ft  \
980   9169 GARLINGTON CT       SACRAMENTO  95829    CA     4      3    2280   
981      6932 RUSKUT WAY       SACRAMENTO  95823    CA     3      2    1477   
982    7933 DAFFODIL WAY   CITRUS HEIGHTS  95610    CA     3      2    1216   
983     8304 RED FOX WAY        ELK GROVE  95758    CA     4      2    1685   
984  3882 YELLOWSTONE LN  EL DORADO HILLS  95762    CA     3      2    1362   

            type                     sale_date   price   latitude   longitude  
980  Residential  Thu May 15 00:00:00 EDT 2008  232425  38.457679 -121.359620  
981  Residential  Thu May 15 00:00:00 EDT 2008  234000  38.499893 -121.458890  
982  Residential  Thu May 15 00:00:00 EDT 2008  235000  38.708824 -121.256803  
983  Residential  Thu May 15 00:00:00 EDT 2008  235301  38.417000 -121.397424  
984  Residential  Thu May 15 00:00:00 EDT 2008  235738  38.655245 -121.075915

### Part 2.2: Import data from JSON file

In a very similar fashion you can load data from a JSON file into a DataFrame. The nice thing with DataFrame - once you have loaded data into it (no matter what the source file format is), you can use the same processing methods.

Your task: load the data from file `colors.json` into a DataFrame called `colors`. 

In [5]:
# Your code here
colors = pd.read_json("Data/colors.json")

Now we print contents of the loaded data:

In [6]:
colors

color value
0      red  #f00
1    green  #0f0
2     blue  #00f
3     cyan  #0ff
4  magenta  #f0f
5   yellow  #ff0
6    black  #000

### Part 2.3: Import data from SQL

Python Pandas allows you to read data from an SQL database into a DataFrame. See function 
[pandas.read_sql_query()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_sql_query.html). 

In this assignment we will use SQLite database, which is a very simple database format, storing everything in a single file. The pandas.read_sql_query() method is very flexible - it takes an SQL query and a connection to SQL interface for any database supporting it. It is not limited to SQLite, you can query MySQL, Oracle and other databases in the same manner. Only the creation of the connection will differ.

Hint: A good (but rather long) tutorial is [here](https://www.dataquest.io/blog/python-pandas-databases/). Search for text "read_sql_query" on the page, you will find a short source code example. 

Your task: load data from the attached file `chinook.sqlite3`. Use the following query to select first five customers, load them into a DataFrame, call it `first_customers`:
```SELECT firstname, lastname, address FROM customers LIMIT 5;```

In [7]:
# Your code here
import sqlite3
conn = sqlite3.connect("Data/chinook.sqlite3")
first_customers = pd.read_sql_query("SELECT firstname, lastname, address FROM customers LIMIT 5;", conn)

Now we print the results:

In [8]:
first_customers

FirstName     LastName                          Address
0       Luís    Gonçalves  Av. Brigadeiro Faria Lima, 2170
1     Leonie       Köhler          Theodor-Heuss-Straße 34
2   François     Tremblay                1498 rue Bélanger
3      Bjørn       Hansen                 Ullevålsveien 14
4  František  Wichterlová                    Klanova 9/506

### Part 2.4: Import data from XML file

This part can be more difficult than loading a CSV. The reason: there is no standard XML reader in Pandas. You will have to implement it yourself. Hint: Google "Pandas DataFrame load XML".

**Your task**: load data from attached file `books.xml`, into variable called `books`.

Hint: you may want to [read the file content into a string variable](https://prefetch.net/blog/index.php/2012/01/30/reading-a-file-into-a-python-string/) first using open() and read() functions.

In [9]:
# Your code here
books = pd.read_xml("Data/books.xml")
books.head()

id                author                  title     genre  price  \
0  bk101  Gambardella, Matthew  XML Developer's Guide  Computer  44.95   
1  bk102            Ralls, Kim          Midnight Rain   Fantasy   5.95   
2  bk103           Corets, Eva        Maeve Ascendant   Fantasy   5.95   
3  bk104           Corets, Eva        Oberon's Legacy   Fantasy   5.95   
4  bk105           Corets, Eva     The Sundered Grail   Fantasy   5.95   

  publish_date                                        description  
0   2000-10-01  An in-depth look at creating applications \n  ...  
1   2000-12-16  A former architect battles corporate zombies, ...  
2   2000-11-17  After the collapse of a nanotechnology \n     ...  
3   2001-03-10  In post-apocalypse England, the mysterious \n ...  
4   2001-09-10  The two daughters of Maeve, half-sisters, \n  ...

## Part 3: Quick data inspection

In this part you do practice simple processing of the loaded data. Python and Pandas is a very rich toolset, here you get just a glimpse of it.

In this part, all tasks are related to the `house` data that you loaded from the CSV file in Part 2. If you are wondering about what methods to use, see [Hints](#Hints) section.

**Task for you**: Select number of rows in the data set, store it in variable `csv_num_rows`.

In [10]:
# Your code here
csv_num_rows = houses.shape[0]
print("Number of rows in the CSV file: %i" % csv_num_rows)

Number of rows in the CSV file: 985


You can select one column of the DataFrame, it will be a Pandas Series object. For example, to select Zip code column:

In [11]:
zipcode = houses["zip"]

**Task for you**: Now select the column "price" into a separate variable called `prices`.

In [12]:
# Your code here
prices = houses.price
prices

0       59222
1       68212
2       68880
3       69307
4       81900
        ...  
980    232425
981    234000
982    235000
983    235301
984    235738
Name: price, Length: 985, dtype: int64

To validate that you are doing correctly, we can use some tests. We know that the 10th price in the file is 98937 (one can simply open the CSV file in a text editor to verify that). Indexing in python starts at zero. It means that prices[9] should be 98937. If you get error after executing the next block of code, something is wrong in your program. There are many possible errors. Therefore before testing the real value, we check if `prices` is a list containing at least 10 values. P.S. The assert statement means "check if this condition is true. If it isn't raise an error".

In [13]:
assert type(prices) == pd.core.series.Series
assert len(prices) >= 10 
assert prices[9] == 98937

Accessing specific rows and items in a DataFrame by providing an index or a range of indices is a topic in itself. We suggest to read a tutorial on it, for example, the [official Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/indexing.html).

**Task for you**: Calculate mean and standard deviation for column "price", store it in variables `price_mean` and `price_std` respectively.

In [14]:
# Your code here
price_mean = sum(prices) / len(prices) #Or prices.mean()
price_std = prices.std()
print(round(price_mean,2), round(price_std,2))

234144.26 138365.84


Pandas DataFrame and Series classes has a handy method `describe()`. We can use it to check whether our calculated mean value is correct:

In [15]:
prices.describe()

count       985.000000
mean     234144.263959
std      138365.839085
min        1551.000000
25%      145000.000000
50%      213750.000000
75%      300000.000000
max      884790.000000
Name: price, dtype: float64

By doing "their homework" the teachers found out that the mean value should be 234144.263959. We make a test to see if your value is the same.

There can be different weird behaviour with number rounding. Therefore, one general hint: never compare floating point numbers on exact values. Instead, you can either round them and compare, or convert them to a string and then compare. In this case, we check if the mean price rounded to integer is 234144.

In [16]:
assert round(price_mean) == 234144

Pandas DataFrame has very rich selection interface. For example, you can easily select all rows, where price equals to 106852. Or all rows where price > 800000. The syntax is a bit weird, just get used to it. :) 
First we create a filter. It will be a list containing True and False values. The i-th value will be True, if the i–th row satisfies the condition. Then we apply that filter to the dataframe and select only the necessary rows: those having True value in the filter. 

In [17]:
# Create a filter - check which rows have price >= 800000
expensive_filter = (houses["price"] >= 800000)

In [18]:
# Now select only those rows in houses DataFrame which satisfiy the filter condition
expensive_houses = houses[expensive_filter]

Now let's print the table showing the expensive houses

In [19]:
expensive_houses

street             city    zip state  beds  baths  sq__ft  \
157          315 JUMEL CT  EL DORADO HILLS  95762    CA     6      5       0   
334      3935 EL MONTE DR           LOOMIS  95650    CA     4      4    1624   
863      2982 ABERDEEN LN  EL DORADO HILLS  95762    CA     4      3       0   
864  9401 BARREL RACER CT           WILTON  95693    CA     4      3    4400   

            type                     sale_date   price   latitude   longitude  
157  Residential  Wed May 21 00:00:00 EDT 2008  830000  38.669931 -121.059580  
334  Residential  Tue May 20 00:00:00 EDT 2008  839000  38.813337 -121.133348  
863  Residential  Fri May 16 00:00:00 EDT 2008  879000  38.706692 -121.058869  
864  Residential  Fri May 16 00:00:00 EDT 2008  884790  38.415298 -121.194858

You can combine the filter and row selection in a single line by putting the filter directly inside the dataframe [] selector:

In [27]:
expensive_houses = houses[houses["price"] >= 800000]

street             city    zip state  beds  baths  sq__ft  \
157          315 JUMEL CT  EL DORADO HILLS  95762    CA     6      5       0   
334      3935 EL MONTE DR           LOOMIS  95650    CA     4      4    1624   
863      2982 ABERDEEN LN  EL DORADO HILLS  95762    CA     4      3       0   
864  9401 BARREL RACER CT           WILTON  95693    CA     4      3    4400   

            type                     sale_date   price   latitude   longitude  
157  Residential  Wed May 21 00:00:00 EDT 2008  830000  38.669931 -121.059580  
334  Residential  Tue May 20 00:00:00 EDT 2008  839000  38.813337 -121.133348  
863  Residential  Fri May 16 00:00:00 EDT 2008  879000  38.706692 -121.058869  
864  Residential  Fri May 16 00:00:00 EDT 2008  884790  38.415298 -121.194858

**Task for you**: Now that you know how to get the statistics (mean, std), how to filter rows and select one column of a DataFrame, do the following:
* Find maximum price in the whole data set
* Find the most expensive house (the one having maximum price)
* Find how many square feet it has and store it in variable `most_expensive_area`

In [38]:
# Your code here
most_expensive_house = houses[houses["price"] == houses["price"].max()].iloc[0]
most_expensive_area = int(most_expensive_house.sq__ft)
print(most_expensive_house)
print(f"Most expensive area: {most_expensive_area}")

street                9401 BARREL RACER CT
city                                WILTON
zip                                  95693
state                                   CA
beds                                     4
baths                                    3
sq__ft                                4400
type                           Residential
sale_date     Fri May 16 00:00:00 EDT 2008
price                               884790
latitude                         38.415298
longitude                      -121.194858
sqft_price                      201.088636
Name: 864, dtype: object
Most expensive area: 4400


Now we test if the value is correct.
Hint: if your result of selection is one cell in Pandas Series, not a number, you can use the .item() method to get the first cell in the Series object.
For example, if you have `area` object of type series, use 

```most_expensive_area = area.item() ```

to get it as a numeric value.

In [39]:
assert most_expensive_area == 4400

You can also do manipulations with all the values in the whole column. Or do an operation on two columns. For example, you can calculate price per square feet and assign it to a new column in the DataFrame:

In [40]:
houses["sqft_price"] = houses["price"] / houses["sq__ft"]

**Task for you**: find the house having the cheapest square meter price. Assign it to variable `cheapest_house`. The resulting variable should be a DataFrame consisting of a single row. Hint: use the selection/filtering techniques described above.

In [50]:
houses["price_per_sqft"] = houses["price"] / houses["sq__ft"]
cheapest_house = houses.loc[[houses["price_per_sqft"].idxmin()]]
cheapest_house

street         city    zip state  beds  baths  sq__ft  \
866  14151 INDIO DR  SLOUGHHOUSE  95683    CA     3      4    5822   

            type                     sale_date  price   latitude   longitude  \
866  Residential  Fri May 16 00:00:00 EDT 2008   2000  38.490447 -121.129337   

     sqft_price  price_per_sqft  
866    0.343525        0.343525

To verify that you found the right house, you can run this code block.

In [51]:
assert type(cheapest_house) == pd.core.frame.DataFrame
assert len(cheapest_house) == 1
assert cheapest_house["price"].item() == 2000 
assert cheapest_house["sq__ft"].item() == 5822

We can use function `pandas.Series.unique()` to get a list of values that a particular series (DataFrame column) takes. Let's take a look what values does the "type" column take:

In [52]:
houses["type"].unique()

array(['Residential', 'Condo', 'Multi-Family', 'Unkown'], dtype=object)

**Task for you**: Find out how many Multi-family houses had less than 3 bathrooms. Store that number in variable `num_few_br_multif`. P.S. by "find out" we mean "write code that finds out". If you find the value by just looking at the file content, it does not count :).

In [54]:
# Your code here
num_few_br_multif = houses[(houses["type"] == "Multi-Family") & (houses["baths"] < 3)].shape[0]
num_few_br_multif

6

**Task for you**: Convert area from square feet to square meters, add it as a column named `area_m2` to the same DataFrame object `houses`.

In [55]:
# Your code here
houses["area_m2"] = houses["sq__ft"] * 0.092903
houses.head()

street        city    zip state  beds  baths  sq__ft  \
0      3526 HIGH ST  SACRAMENTO  95838    CA     2      1     836   
1       51 OMAHA CT  SACRAMENTO  95823    CA     3      1    1167   
2    2796 BRANCH ST  SACRAMENTO  95815    CA     2      1     796   
3  2805 JANETTE WAY  SACRAMENTO  95815    CA     2      1     852   
4   6001 MCMAHON DR  SACRAMENTO  95824    CA     2      1     797   

          type                     sale_date  price   latitude   longitude  \
0  Residential  Wed May 21 00:00:00 EDT 2008  59222  38.631913 -121.434879   
1  Residential  Wed May 21 00:00:00 EDT 2008  68212  38.478902 -121.431028   
2  Residential  Wed May 21 00:00:00 EDT 2008  68880  38.618305 -121.443839   
3  Residential  Wed May 21 00:00:00 EDT 2008  69307  38.616835 -121.439146   
4  Residential  Wed May 21 00:00:00 EDT 2008  81900  38.519470 -121.435768   

   sqft_price  price_per_sqft     area_m2  
0   70.839713       70.839713   77.666908  
1   58.450728       58.450728  108.417801  
2   86.532663       86.532663   73.950788  
3   81.346244       81.346244   79.153356  
4  102.760351      102.760351   74.043691

**Task for you**: A bit more challenging exercise: select the top five cities by the total area of real estate sold. Show them order by the area, in descending order. I.e, the result should be a Pandas Series with five rows where the index contains city names and the value column contains total area sold in that city. Store result in variable `top_five_cities`.

Hint: If you do the calculations correctly, the top #3 city should be Antelope with total area sold ~ 5262.86 m2.

In [59]:
# Your code here
top_five_cities = (houses.groupby("city")["area_m2"].sum().sort_values(ascending=False).head(5))
top_five_cities

city
SACRAMENTO        56864.068240
ELK GROVE         19579.771765
ANTELOPE           5262.862047
ROSEVILLE          4587.271431
CITRUS HEIGHTS     4255.329012
Name: area_m2, dtype: float64

## Reflection

Please reflect on the following questions:
1. How did the assignment go? Was it easy or hard?
    
    -The assignment was easy in the beginning but got progressivly harder but i managed with some research and help.
2. How many hours did you spend on it?
    
    -I would say around 3 hours
3. What was the most time-consuming part?
    
    -The last few because i needed to reseach alot and use hints.
4. If you need to do similar things later in your professional life, how can you improve? How can you do it more efficiently?
   
    -I do not know how to do this more effective than above. Maby you could learn the library better.
5. Was there something you would expect to learn that this exercise did not include?
   
    -No.
6. Was there something that does not make sense?
   
    -No.

**Your answers here**

## Hints

The following Pandas functions may be handy
* [read_csv in Pandas](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) may be useful. There are lots of possible parameters to the function. You probably need only the parameter specifying filename.
* [DataFrame.head()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.head.html) and [DataFrame.tail()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.tail.html).
* Series.mean(), .std() and .max() functions for statistics
* [DataFrame.groupby()](https://pandas.pydata.org/pandas-docs/stable/groupby.html) for grouping rows together; [.sort_values()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html) for sorting the rows by calues in specific columns.

A good [tutorial with a generic XML reading example](http://www.austintaylor.io/lxml/python/pandas/xml/dataframe/2016/07/08/convert-xml-to-pandas-dataframe/) is provided by Austin Taylor. 

It is quite common for programmers to not remember all the functions and all possible parameters. One skill that is important to master: ability to quickly find documentation and examples on the internet. Several tips:
1. Try typing your question in Google. For example, if you wonder how to read a CSV file in this context, you can try to Google ["python Pandas read CSV DataFrame"](https://www.google.com/?q=python%20Pandas%20read%20CSV%20Dataframe) and with high chance one of the top 5 results will give you the answer. Just "Read CSV" will not be enough. You should mention the context. For the assignments of this course the context typically is "Python Pandas". 
2. [Stackoverflow](https://stackoverflow.com/) is one of the best resources for programming questions. It has a lot of different questions and answers. When you have a question, there is a 90% chance that someone has had exactly that question before and someone else has answered him/her. But typically it is enough to type the question in Google. All roads (i.e., programming questions) lead to StackOverflow anyway :).
3. Check out the official documentation of the libraries that you are using. If you wonder about some DataFrame functions, find the DataFrame page in Pandas documentation. Hint: typically Google is smart enough to bring you to the documentation page. If you simply type ["Pandas DataFrame"](https://www.google.no/search?q=pandas+dataframe&oq=pandas+dataframe), the first result will typically be the documentation page. Official documentation is typically the page with the most detailed information on exactly the class or function that you need to explore. If you wonder about the possible function parameters, etc, see the docs.